x/sync/manager.go

```diff
@@ -510,43 +541,113 @@ func (m *Manager) handleRangeProofResponse(
 		return err
 	}
 
-	var rangeProofProto pb.RangeProof
-	if err := proto.Unmarshal(responseBytes, &rangeProofProto); err != nil {
+	rangeProof, err := m.proofParser.RangeProof(
+		responseBytes,
+		request.RootHash,
+		maybeBytesToMaybe(request.StartKey),
+		maybeBytesToMaybe(request.EndKey),
+		request.KeyLimit,
+	)
+	if err != nil {
 		return err
 	}
 
-	var rangeProof merkledb.RangeProof
-	if err := rangeProof.UnmarshalProto(&rangeProofProto); err != nil {
+	if err := rangeProof.Verify(ctx); err != nil {
 		return err
 	}
 
-	if err := verifyRangeProof(
-		ctx,
-		&rangeProof,
-		int(request.KeyLimit),
-		maybeBytesToMaybe(request.StartKey),
-		maybeBytesToMaybe(request.EndKey),
-		request.RootHash,
-		m.tokenSize,
-		m.config.Hasher,
-	); err != nil {
-		return err
+	// Replace all the key-value pairs in the DB from start to end with values from the response.
+	if err := rangeProof.Commit(ctx); err != nil {
+		m.setError(err)
+		return nil
 	}
 
-	largestHandledKey := work.end
-
-	// Replace all the key-value pairs in the DB from start to end with values from the response.
-	if err := m.config.DB.CommitRangeProof(ctx, work.start, work.end, &rangeProof); err != nil {
+	nextKey, err := rangeProof.FindNextKey(ctx)
+	if err != nil {
 		m.setError(err)
 		return nil
 	}
+	m.completeWorkItem(work, nextKey, targetRootID)
+	return nil
+}
+
+func (p *rangeProof) Verify(ctx context.Context) error {
+	return verifyRangeProof(
+		ctx,
+		p.merkleProof,
+		p.request.keyLimit,
+		p.request.startKey,
+		p.request.endKey,
+		p.request.rootHash,
+		p.tokenSize,
+		p.hasher,
+	)
+}
+
+func (p *rangeProof) Commit(ctx context.Context) error {
```

Comment:
nit: idiomatic in go to have a receiver starting with the first letter of the type, so `r` would be idiomatic here.